app/models/user_auth_token.rb

```diff
@@ -30,23 +30,23 @@ def self.log(info)
     end
   end
 
-  # Returns the login location as it will be used by the the system to detect
-  # suspicious login.
-  #
-  # This should not be very specific because small variations in location
-  # (i.e. changes of network, small trips, etc) will be detected as suspicious
-  # logins.
-  #
-  # On the other hand, if this is too broad it will not report any suspicious
-  # logins at all.
-  #
-  # For example, let's choose the country as the only component in login
-  # locations. In general, this should be a pretty good choce with the
-  # exception that for users from huge countries it might not be specific
-  # enoguh. For US users where the real user and the malicious one could
-  # happen to live both in USA, this will not detect any suspicious activity.
+  MAX_SUSPICIOUS_DISTANCE_KM = 500 # kilometers
+  RAD_PER_DEG = Math::PI / 180
+  EARTH_RADIUS_KM = 6371 # kilometers
+
   def self.login_location(ip)
-    DiscourseIpInfo.get(ip)[:country]
+    ipinfo = DiscourseIpInfo.get(ip)
+    [ipinfo['latitude'], ipinfo['longitude']]
+  end
+
+  def self.distance(loc1, loc2)
+    lat1_rad, lon1_rad = loc1[0] * RAD_PER_DEG, loc1[1] * RAD_PER_DEG
+    lat2_rad, lon2_rad = loc2[0] * RAD_PER_DEG, loc2[1] * RAD_PER_DEG
+
+    a = Math.sin((lat2_rad - lat1_rad) / 2)**2 + Math.cos(lat1_rad) * Math.cos(lat2_rad) * Math.sin((lon2_rad - lon1_rad) / 2)**2
+    c = 2 * Math::atan2(Math::sqrt(a), Math::sqrt(1 - a))
+
+    c * EARTH_RADIUS_KM
```

Comment:
I think we should improve the names of the variables used for `a` and `c`.

Reply:
They do not have any particular meaning and are just partial results of the formula. I have checked now and most implementations I could find on the internet just mention them as `a` and `c`. 🤔

Reply:
ah icic. This is fine then :+1: